server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,383 @@
+                active = false;
+                handleWakeUp();
+            }
+            this.leader = Optional.of(leader);
+            assert assertInactiveWithNoKnownPeers();
+        }
+    }
+
+    // exposed to subclasses for testing
+    protected final boolean holdsLock() {
+        return Thread.holdsLock(mutex);
+    }
+
+    boolean assertInactiveWithNoKnownPeers() {
+        assert active == false;
+        assert peersByAddress.isEmpty();
+        return true;
+    }
+
+    PeersResponse handlePeersRequest(PeersRequest peersRequest) {
+        synchronized (mutex) {
+            assert peersRequest.getSourceNode().equals(getLocalNode()) == false;
+            if (active) {
+                startProbe(peersRequest.getSourceNode().getAddress());
+                peersRequest.getKnownPeers().stream().map(DiscoveryNode::getAddress).forEach(this::startProbe);
+                return new PeersResponse(Optional.empty(), getKnownPeers(), currentTerm);
+            } else {
+                return new PeersResponse(leader, Collections.emptyList(), currentTerm);
+            }
+        }
+    }
+
+    public void setCurrentTerm(long currentTerm) {
+        this.currentTerm = currentTerm;
+    }
+
+    private DiscoveryNode getLocalNode() {
+        final DiscoveryNode localNode = transportService.getLocalNode();
+        assert localNode != null;
+        return localNode;
+    }
+
+    /**
+     * Called on receipt of a PeersResponse from a node that believes it's an active leader, which this node should therefore try and join.
+     */
+    protected abstract void onActiveMasterFound(DiscoveryNode masterNode, long term);
+
+    public interface TransportAddressConnector {
+        /**
+         * Identify the node at the given address and, if it is a master node and not the local node then establish a full connection to it.
+         */
+        void connectToRemoteMasterNode(TransportAddress transportAddress, ActionListener<DiscoveryNode> listener);
+    }
+
+    private List<DiscoveryNode> getKnownPeers() {
+        assert active;
+        assert holdsLock() : "PeerFinder mutex not held";
+        List<DiscoveryNode> knownPeers = new ArrayList<>(peersByAddress.size());
+        for (final Peer peer : peersByAddress.values()) {
+            DiscoveryNode peerNode = peer.getDiscoveryNode();
+            if (peerNode != null) {
+                knownPeers.add(peerNode);
+            }
+        }
+        return knownPeers;
+    }
+
+    private Peer createConnectingPeer(TransportAddress transportAddress) {
+        Peer peer = new Peer(transportAddress);
+        peer.establishConnection();
+        return peer;
+    }
+
+    private void handleWakeUp() {
+        assert holdsLock() : "PeerFinder mutex not held";
+
+        for (final Peer peer : peersByAddress.values()) {
+            peer.handleWakeUp();
+        }
+
+        if (active == false) {
+            logger.trace("ActivePeerFinder#handleWakeUp(): not running");
```

Comment:
rename to PeerFinder

Reply:
Gotta catch them all. See 2051db6363d1b5105ab578b3ce390e4c655d44ee.